rust-peer/src/main.rs

```diff
@@ -102,13 +106,43 @@ async fn main() -> Result<()> {
                 )) => {
                     info!("{peer_id} subscribed to {topic}");
                 }
+                SwarmEvent::Behaviour(BehaviourEvent::Identify(e)) => {
+                    debug!("{:?}", e);
+
+                    if let identify::Event::Received {
+                        peer_id,
+                        info:
+                            identify::Info {
+                                listen_addrs,
+                                protocols,
+                                observed_addr,
+                                ..
+                            },
+                    } = e
+                    {
+                        swarm.add_external_address(observed_addr, AddressScore::Infinite);
+
+                        if protocols
+                            .iter()
+                            .any(|p| p.as_bytes() == KADEMLIA_PROTOCOL_NAME)
+                        {
+                            for addr in listen_addrs {
+                                swarm.behaviour_mut().kademlia.add_address(&peer_id, addr);
+                            }
+                        }
+                    }
+                }
                 event => {
                     debug!("{event:?}");
                 }
             },
             futures::future::Either::Right(_) => {
                 tick = futures_timer::Delay::new(TICK_INTERVAL);
 
+                if let Err(e) = swarm.behaviour_mut().kademlia.bootstrap() {
+                    debug!("Failed to run Kademlia bootstrap: {e:?}");
```

Comment:
This should probably be logged as an error.

```suggestion
                    error!("Failed to run Kademlia bootstrap: {e:?}");
```